File: perl/dancer/geotz.pl
use Dancer;

use Try::Tiny;
use Regexp::Common qw(net);
use DateTime;
use DateTime::TimeZone;
use Geo::IP;

set logger => 'console';

my $gi = Geo::IP->open( "/usr/local/share/GeoIP/GeoLiteCity.dat", GEOIP_MEMORY_CACHE ) or die;

post '/' => sub {
    my $ref;
    try {
        $ref = from_json(request->body, { allow_nonref => 1 });
    }
    catch {
        send_error('Bad request', 400);
    };

    if ( ref($ref) ne "ARRAY" ) {
        $ref = [ $ref ];
    }

    my $out;
    foreach my $ip ( @{ $ref } ) {
        # Skip if this value isn't a dotted quad (e.g., "1.1.1.1")
        next if $ip !~ /^$RE{net}{IPv4}{dec}$/;
        my $record = $gi->record_by_addr($ip);
        my $country = $record->country_name;
        my $region = $record->region_name;
        $out->{$ip} = {
            country => $country,
            region => $region,
        };
        my $tz = $record->time_zone;
        if ( $tz ) {
            my $dt = DateTime->now();
            my $dtz = DateTime::TimeZone->new( name => $tz );
            $out->{$ip}->{timezone} = $tz;
            $out->{$ip}->{utc_offset} = $dtz->offset_for_datetime($dt),
        }
        else {
            $out->{$ip}->{timezone} = 'unknown';
            $out->{$ip}->{utc_offset} = 0;
        }

    }

    content_type 'application/json';
    return to_json($out);
};

start;

File: dotcloud/flask/geotz/geotz.py
from flask import Flask, request, make_response
from pytz import timezone
from datetime import datetime
import json
import pygeoip

DEBUG=True

app = Flask(__name__)
app.config.from_object(__name__)

gi = pygeoip.GeoIP("GeoLiteCity.dat")

@app.route('/', methods=['POST'])
def lookup():
    rv = dict()
    for ip in request.json:
        r = gi.record_by_addr(ip)
        tz = timezone(r['time_zone'])
        rv[ip] = { 
            'country': r['country_name'],
#            'region': r['region'],
            'city': r['city'],
            'time_zone': r['time_zone'],
            'utc_offset': int(tz.utcoffset(datetime.utcnow()).total_seconds())
        }

    response = make_response(json.dumps(rv))
    response.headers['Content-Type'] = 'application/json'
    return response

if __name__ == '__main__':
    app.run()

